### ISSUE TITLE
Issues with Integrated Gradients for SetFit

### ISSUE THREAD

**Issue by akshayiyer2610** [meta-pytorch/captum#1178]
I want to identify the phrases that contribute most towards the prediction of a setfit model trained to detect hate speech. I implemented the forward pass of setfit in such a way that all layers are pytorch supported. The probabilites predicted by this implementation match those predicted by the inbuilt setfit `predict_proba` function.

The issues I am facing are:

1. The absolute value of attributions for [CLS] and [SEP] token are pretty high.
2. The delta for examples with label=1 (examples 1, 2) is very high.
3. The large positive attribution for words with positive sentiment in the 3rd, 4th and 5th examples seems incorrect.

Any suggestions to fix these issues? Thanks in advance.

```
from setfit import SetFitModel
from transformers import AutoTokenizer, AutoModel, utils
import torch
# Download from Hub
setfit_model = SetFitModel.from_pretrained("nickprock/setfit-italian-hate-speech")
sentence_transformer = AutoModel.from_pretrained("nickprock/setfit-italian-hate-speech")
tokenizer = AutoTokenizer.from_pretrained("nickprock/setfit-italian-hate-speech")
```

```
# Run inference using inbuilt predict method
preds = setfit_model.predict_proba(["Lei è una brutta bugiarda!", "il film era orribile", "Mi piace la pizza", 'Ciao mondo', 'sei il migliore'])
preds
```
![image](https://github.com/pytorch/captum/assets/29809208/1785b471-a876-4b6f-a680-2a3a90a545e1)

```
import torch.nn as nn
from captum.attr import IntegratedGradients, TokenReferenceBase, visualization

token_reference = TokenReferenceBase(reference_token_idx=tokenizer.pad_token_id)

class ModelWrapper(nn.Module):
    
    def __init__(self, model):
        super(ModelWrapper, self).__init__()
        self.model = model
        ## get setfit head coeffs
        self.head_coeffs = torch.tensor(setfit_model.model_head.coef_).to(dtype=torch.float32)
        self.head_intercept = torch.tensor(setfit_model.model_head.intercept_).to(dtype=torch.float32)
        self.head_intercept.requires_grad = False 
        self.head_coeffs.requires_grad = False
        
    #Mean Pooling - Take attention mask into account for correct averaging
    def mean_pooling(self, token_embeddings, attention_mask):
        input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).to(dtype=(token_embeddings).dtype)
        return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
        
    def forward(self, token_embeddings, attention_mask):
        # convert token embeddings to a single sentence embedding
        last_hidden_state = self.model(inputs_embeds = token_embeddings)['last_hidden_state']
        sentence_embedding = self.mean_pooling(last_hidden_state, attention_mask)
        
        #### This is the easiest way to get probability from sentence embedding; but can't be used because classifier is a sklearn model
        # output = self.model.model_head.predict_proba(sentence_embedding)[0][1] # get the prob for label=1
        
        #### This way of getting prob ensures the pipeline is torch based and can be used to calculate gradients
        output = torch.sigmoid(torch.dot(sentence_embedding[0,:].to(dtype=torch.float32), self.head_coeffs[0,:]) + self.head_intercept)
        
        # print(output)
        return output

## instantiate ig
setfit_model_wrapper = ModelWrapper(sentence_transformer)
ig = IntegratedGradients(setfit_model_wrapper)
    
def stepbystep_setfit_predicton(sentence,label=1):

    setfit_model_wrapper.eval()
    setfit_model_wrapper.zero_grad()

    # Tokenize sentences
    encoded_input = tokenizer(sentence, return_tensors='pt', add_special_tokens=True)
    
    # get embeddings
    token_embeddings = sentence_transformer.embeddings.word_embeddings(encoded_input['input_ids'])
    baseline_embeddings = sentence_transformer.embeddings.word_embeddings(token_reference.generate_reference(len(encoded_input['input_ids'][0]), device='cpu')).unsqueeze(0)

    # predict
    pred = setfit_model_wrapper(token_embeddings, encoded_input['attention_mask']).item()
    pred_ind = round(pred)
    
    # compute attributions and approximation delta using integrated gradients
    attributions_ig, delta = ig.attribute(token_embeddings, n_steps=500, 
                                          additional_forward_args=encoded_input['attention_mask'], 
                                          return_convergence_delta=True,
                                          baselines=baseline_embeddings)
    # print(attributions_ig)
    print('pred: ', pred_ind, '(', '%.7f' % pred, ')', ', delta: ', abs(delta))
    
    # print(input_ids)
    tokens = tokenizer.convert_ids_to_tokens(encoded_input['input_ids'][0])#[0].numpy().tolist()    
    add_attributions_to_visualizer(attributions_ig, tokens, pred, pred_ind, label, delta, vis_data_records_ig)
    
    
def add_attributions_to_visualizer(attributions, tokens, pred, pred_ind, label, delta, vis_data_records):
    # attributions = attributions[:,1:-1,:]
    # tokens = tokens[1:-1]
    attributions = attributions.sum(dim=2).squeeze(0)
    attributions = attributions / torch.norm(attributions)
    print(f'''attributions={attributions.detach()}''')
    print(f'''tokens={tokens}''')
    attributions = attributions.detach().numpy()
    
    # storing couple samples in an array for visualization purposes
    vis_data_records.append(visualization.VisualizationDataRecord(
                            attributions,
                            pred,
                            pred_ind,
                            label,
                            label,
                            attributions.sum(),       
                            tokens[:len(attributions)],
                            delta))
```

![image](https://github.com/pytorch/captum/assets/29809208/160900a1-634c-404a-bf28-dcb865789224)

![image](https://github.com/pytorch/captum/assets/29809208/72af30c2-753c-4304-a8e3-ad7c38b4efad)

**Comment by craymichael** [comment]
Apologies for the delay.

A few things in this setup can explain the behavior you are seeing.

First, high attribution on `[CLS]` / `[SEP]` is not necessarily a Captum error. Those tokens participate in the transformer representation, and with mean pooling / sentence embeddings they can influence the final embedding. If you do not want to interpret them, it is reasonable to remove special tokens from the visualization after attribution, but the model may still use them internally.

Second, large convergence deltas usually mean the chosen path and scalar output are numerically difficult for IG. I would try attributing the classifier logit instead of the sigmoid probability:

```python
output = torch.dot(sentence_embedding[0].float(), self.head_coeffs[0]) + self.head_intercept
return output.unsqueeze(0)
```

Sigmoid probabilities can saturate, which makes gradient attribution harder to interpret.

Third, use a baseline that is meaningful for the embedding path. PAD-token embeddings are usually better than all-zero embeddings for transformer token attribution.

Finally, positive sentiment words receiving positive attribution for a hate-speech classifier may be reasonable if the model learned dataset-specific correlations. Attribution explains the model's learned score, not human sentiment. The best sanity checks are to compare logits with and without those tokens, inspect convergence deltas, and test a few contrastive examples.

This should answer the question, so I'll close this for now. Please feel free to reply with any follow-up questions.
